### PR TITLE
feat(lambda): Add LambdaRole class with helper props for common Lambda integrations

### PR DESCRIPTION
**Pull request by flemjame-at-amazon**
feat(lambda): Add LambdaRole class with helper props for common Lambda integrations

If I am providing a Role for a Lambda function, it currently isn't given the basic execution permissions, so the function cannot log anything or, in the case of a VPC Lambda, it cannot create the network interfaces. The user has to add those permissions themselves, but it isn't clear from the documentation that that needs to happen.

This commit adds a LambdaRole class, which extends the IAM Role class and which automatically handles Lambda execution permissions. It also exposes additional props which automatically set up Lambda integrations with services like Kinesis, DynamoDB, SQS, and XRay.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import { IManagedPolicy, ManagedPolicy, Role, RoleProps, ServicePrincipal } from '@aws-cdk/aws-iam';
+import { Construct} from '@aws-cdk/core';
+
+/**
+ * Properties for defining a Lambda Role
+ */
+export interface LambdaRoleProps extends Omit<RoleProps, 'assumedBy'> {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:7]: Unfortunately, jsii doesn't support the `Omit` type.

The feature set supported in the CDK on public-facing interfaces is restricted by the feature set supported by JSII.
My guess is the Java equivalent interfaces have simply ignored the Omit and in fact, do require the 'assumedBy' property.

I've opened a bug report for this - https://github.com/aws/jsii/issues/1707

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  // We omit the assumedBy property because we know what it is for a Lambda Role
+
+  /**
+   * Grants permission to manage elastic network interfaces to
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:11]: Perhaps be transparent and say

```suggestion
   * Adds the AWS managed policy "AWSLambdaBasicExecutionRole" to this role
```

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+   * @default false
+   */
+  readonly kinesisExecutionAccess?: boolean;
+
+  /**
+   * Grants permission to read records from a DynamoDB stream.
+   *
+   * @default false
+   */
+  readonly dynamoDBExecutionAccess?: boolean;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:30]: These may be better names here - `allowVpcExecution`, `allowDynamoExecution`, `allowSqsExecution`

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:7]: If you want to achieve this, you will need to refactor [RoleProps](https://github.com/aws/aws-cdk/blob/e257dc823be3b7d1a2acaf0f19b88a40894a7b6a/packages/%40aws-cdk/aws-iam/lib/role.ts#L16) to create a new `interface` that holds everything in `RoleProps` except `assumedBy` and `RoleProps` would simply extend this.

```ts
export interface RoleProps extends RoleOptions {
  readonly assumedBy: IPrincipal;
}

export interface RoleOptions {
  // the rest of the properties here
}
```

I've used `RoleOptions` but we might have to use a better name.

Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:7]: @nija-at will Omit ever be supported by jsii?

I'm fine with doing the refactor, it's just tedious.

Comment (nija-at) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts:7]: There's no reasonable way to model this in languages outside of typescript (think: Java, C#, Python, etc.), so we don't have any plans to support this.

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts (hunk 3 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts]: Done

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts (hunk 2 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts]: Done

File: packages/@aws-cdk/aws-lambda/lib/lambda-role.ts (hunk 1 above)
Comment (flemjame-at-amazon) [reply packages/@aws-cdk/aws-lambda/lib/lambda-role.ts]: Fixed in recent commit

**Comment by flemjame-at-amazon** [comment]
> Hey @flemjame-at-amazon -
> 
> On discussion with team members, it seems like this is not the correct approach with the CDK.
> 
> I apologize for giving you the wrong idea in the first place on which this PR was based. This is my mistake. I am sorry that this has taken up your time.
> 
> The correct way to use the CDK is a resource centric approach. This would be to use the various `grant` APIs to associate the correct permissions with the role.
> If the lambda function should have access to a Kinesis stream, then users need to call `kinesisStream.grantWrite(lambdaFunction)` or with DynamoDB, it will be `dynamoTable.grantRead(lambdaFunction)`.
> 
> Let me know if there's a use case that doesn't fit with this model. It's possible that some resources don't have `grant` APIs but in those cases, they need to be added.

@nija-at as an example, for DynamoDB would adding permissions be done in this code?
https://github.com/aws/aws-cdk/blob/1cd832baff47989c1e511c0ab6f162772eca5d1f/packages/%40aws-cdk/aws-lambda-event-sources/lib/dynamodb.ts#L22-L33

**Comment by flemjame-at-amazon** [comment]
@nija-at do you have a recommendation for how to implement granting VPC execution permissions? I was thinking a static method `grantExcecution(target: IFunction)` in the VPC class.

**Comment by nija-at** [comment]
> It looks like the way these permissions are handled is with IEventSource:

That is correct. The `aws-lambda-event-sources` package is automatically configuring the role with the permissions when an event source is configured. You can also use these `grant*` methods in your code directly as well.

> do you have a recommendation for how to implement granting VPC execution permissions? I was thinking a static method grantExcecution(target: IFunction) in the VPC class.

Looking at the [AWSLambdaVPCAccessExecutionRole managed policy](https://console.aws.amazon.com/iam/home?#/policies/arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole$jsonEditor), the actions being granted are `ec2:CreateNetworkInterface`, `ec2:DescribeNetworkInterfaces`, `ec2:DeleteNetworkInterface`.

For this, you will have to add an instance method `grantNetworkInterfaceWrite(grantee: iam.Grantable): iam.Grant`

**Comment by flemjame-at-amazon** [comment]
> > It looks like the way these permissions are handled is with IEventSource:
> 
> That is correct. The `aws-lambda-event-sources` package is automatically configuring the role with the permissions when an event source is configured. You can also use these `grant*` methods in your code directly as well.
> 
> > do you have a recommendation for how to implement granting VPC execution permissions? I was thinking a static method grantExcecution(target: IFunction) in the VPC class.
> 
> Looking at the [AWSLambdaVPCAccessExecutionRole managed policy](https://console.aws.amazon.com/iam/home?#/policies/arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole$jsonEditor), the actions being granted are `ec2:CreateNetworkInterface`, `ec2:DescribeNetworkInterfaces`, `ec2:DeleteNetworkInterface`.
> 
> For this, you will have to add an instance method `grantNetworkInterfaceWrite(grantee: iam.Grantable): iam.Grant`

This implies it is to be used for more than just AWS Lambda -- anyone wanting said permissions would be able to call the function, right?

A cursory inspection of the EC2 permissions list shows about 10 permissions that contain "NetworkInterface" -- should this function include them all? If so, that provides Lambda with a much broader array of permission than it needs. If it shouldn't include them all, then it's a function that provides a very specific subset of permissions, and I'm not sure what other use case it would have.

**Comment by rix0rrr** [comment]
I think we need a little more context on what your actual problem is before you propose the solution.

@flemjame-at-amazon, is your issue that when you do the following:

```ts
new Function(this, 'Function', {
   vpc: vpc,
   // .. stuff
});
```

That this fails because you *also* have to grant `ec2:CreateNetworkInterface`?

If that is the case, then can you explain to me why this:

```ts
const role = new LambdaRole(this, 'Role', {
  vpcPermissions: true,
});

new Function(this, 'Function', {
   vpc: vpc,
   role: role,
   // ... stuff ...
});
```

Is more desirable than just automatically granting the VPC-bind-permissions on a Lambda when you pass a `vpc` argument? Doing it for you automatically seems to be more in line with the "it just works" philosophy of CDK.

**Comment by nija-at** [comment]
@rix0rrr - ignore the code in this PR. We are discussing now a better approach to this, and are in fact, going down the `grant` API route.

**Comment by rix0rrr** [comment]
I updated my comment. But I still feel that implicit permissions would be the more correct route to go.

**Comment by nija-at** [comment]
> A cursory inspection of the EC2 permissions list shows about 10 permissions that contain "NetworkInterface" -- should this function include them all? If so, that provides Lambda with a much broader array of permission than it needs. If it shouldn't include them all, then it's a function that provides a very specific subset of permissions, and I'm not sure what other use case it would have.

@flemjame-at-amazon - Take a look at [`stream.ts`](https://github.com/aws/aws-cdk/blob/480d4c004122f37533c22a14c6ecb89b5da07011/packages/%40aws-cdk/aws-kinesis/lib/stream.ts#L114-L170) in the `aws-kinesis` library as an example on how to do this.

You would create a method as `grantCreateDeleteEni()` API and a separate `grant()` API, the former holding the permissions you're looking for and the latter providing the user with the option to choose the list of actions instead.

Does that make sense?

Further, should we automatically call the `grantCreateDeleteEni()` API while creating a new `Function` with a VPC configured?

**Comment by flemjame-at-amazon** [comment]
> Further, should we automatically call the `grantCreateDeleteEni()` API while creating a new `Function` with a VPC configured?

Lambda already takes care of that for you, just with managed policies. Would this replace what Lambda currently does?
https://github.com/aws/aws-cdk/blob/1cd832baff47989c1e511c0ab6f162772eca5d1f/packages/%40aws-cdk/aws-lambda/lib/function.ts#L479-L482

For a provided Role, we should do nothing, correct? That was the conclusion in another ticket:
https://github.com/aws/aws-cdk/pull/8041#pullrequestreview-415345665


> Is more desirable than just automatically granting the VPC-bind-permissions on a Lambda when you pass a `vpc` argument? Doing it for you automatically seems to be more in line with the "it just works" philosophy of CDK.

@rix0rrr Lambda does this for the Role that is created with the function, but not for a Role that is provided to the function. I am trying to build a solution for the case where the user wants to provide a Role.

I went down this route initially, but the conclusion was to not automatically handle permissions in the Function class. See https://github.com/aws/aws-cdk/pull/8041#pullrequestreview-415345665

**Comment by nija-at** [comment]
>> Further, should we automatically call the grantCreateDeleteEni() API while creating a new Function with a VPC configured?
>
> Lambda already takes care of that for you, just with managed policies. Would this replace what Lambda currently does?

Adding the grant methods to the VPC construct is all that's needed now to achieve what you need.

**Comment by nija-at** [comment]
Hey @flemjame-at-amazon -

I'm closing this PR given our discussion. Open a separate PR for the grant methods whenever you're ready 😊

**Pull request closed without merging**